cmd/lncli/cmd_pay.go

```diff
@@ -334,54 +334,73 @@ func sendPaymentRequest(ctx *cli.Context, req *lnrpc.SendRequest) error {
 		}
 	}
 
-	amt := req.Amt
-
+	var feeLimit int64
 	if req.PaymentRequest != "" {
-		req := &lnrpc.PayReqString{PayReq: req.PaymentRequest}
-		resp, err := client.DecodePayReq(context.Background(), req)
+		// Decode payment request to find out the amount.
+		decodeReq := &lnrpc.PayReqString{PayReq: req.PaymentRequest}
+		decodeResp, err := client.DecodePayReq(
+			context.Background(), decodeReq,
+		)
 		if err != nil {
 			return err
 		}
 
-		invoiceAmt := resp.GetNumSatoshis()
+		// If amount is present in the request, override the request
+		// amount.
+		amt := req.Amt
+		invoiceAmt := decodeResp.GetNumSatoshis()
 		if invoiceAmt != 0 {
 			amt = invoiceAmt
 		}
 
+		// Calculate fee limit based on the determined amount.
+		feeLimit, err = retrieveFeeLimit(ctx, amt)
+		if err != nil {
+			return err
+		}
+
+		// Ask for confirmation of amount and fee limit if payment is
+		// forced.
 		if !ctx.Bool("force") {
-			err := confirmPayReq(resp, amt)
+			err := confirmPayReq(decodeResp, amt, feeLimit)
 			if err != nil {
 				return err
 			}
 		}
+	} else {
+		var err error
+		feeLimit, err = retrieveFeeLimit(ctx, req.Amt)
+		if err != nil {
+			return err
+		}
 	}
 
-	paymentStream, err := client.SendPayment(context.Background())
-	if err != nil {
-		return err
-	}
-
-	if err := paymentStream.Send(req); err != nil {
-		return err
-	}
+	req.FeeLimitSat = feeLimit
 
-	resp, err := paymentStream.Recv()
+	stream, err := routerClient.SendPayment(context.Background(), req)
 	if err != nil {
 		return err
 	}
 
-	paymentStream.CloseSend()
+	for {
+		status, err := stream.Recv()
+		if err != nil {
+			return err
+		}
 
-	printRespJSON(resp)
+		if status.State != routerrpc.PaymentState_IN_FLIGHT {
+			printRespJSON(status)
 
-	// If we get a payment error back, we pass an error
-	// up to main which eventually calls fatal() and returns
-	// with a non-zero exit code.
-	if resp.PaymentError != "" {
-		return errors.New(resp.PaymentError)
-	}
+			// If we get a payment error back, we pass an error up
+			// to main which eventually calls fatal() and returns
+			// with a non-zero exit code.
+			if status.State != routerrpc.PaymentState_SUCCEEDED {
```

Comment:
It's a bit hard to follow but won't this print the error twice? (once in `printRespJSON` and then wherever we process the error from this call)

Reply:
Yes, that is indeed the case. I could also return a more generic error like 'payment failed', to not duplicate the detailed reason. But if we are printing an error line anyway, maybe it is not bad to give the detail as well?

Reply:
I'm OK with this if this is OK, just wanted to take notice here.